visn/src/lib.rs

```diff
@@ -57,71 +57,127 @@ use std::collections::VecDeque;
 /// each test needs only define a starting state, a list of events to execute, and
 /// an expected ending condition.
 ///
-
-pub trait KnowledgeEngine<S, E>: Sized {
-    fn queue_event(self, event: E) -> Self;
-    fn queue_events(self, events: &[E]) -> Self;
-    fn resolve_with<F: FnOnce(&mut dyn Iterator<Item = E>) -> &mut dyn Iterator<Item = E>>(
+/// # Types
+/// - `System`: the type of the system under test.
+/// - `Event`: the type of synthetic events.
+/// - `Return`: the type returned by the `resolve` function. Can be the same as `S`, or
```

Comment:
Still referencing `S` here.

Btw, I'm not against using not-single letter type names here, but am curious why you wanted to spell them out. `S`, `E` and `R` wouldn't have been _too_ obscure, probably

Reply:
I personally found them really confusing while doing refactorings but that might not impact the end user as much, I'm not sure.

Reply:
To expand on that: `S`, `E`, `R` aren't to bad by themselves, but with `F` and `G` for function types and `P` for the concrete internal success type for `FallibleEngineImpl`, it gets confusing.